### PR TITLE
Add merged json_encoders inheritance

### LINKED ISSUE #2024: json_encoders inheritance
# Feature Request

Currently, defining `json_encoders` in child classes completely overrides the one defined from the parent classes:
```py
from datetime import datetime, timedelta
from pydantic import BaseModel

class A(BaseModel):
    dt: datetime = datetime.now()
    class Config:
        json_encoders = {datetime: lambda _: "encoded_from_A"}

print(A().json())
#> {"dt": "encoded_from_A"}
class B(A):
    timedt: timedelta = timedelta(hours=100)
    class Config:
        json_encoders = {timedelta: lambda _: "encoded_from_B"}

print(B().json())
#> {"dt": "2020-10-23T01:11:20.613624", "timedt": "encoded_from_B"}

```

I think it would be handy to merge the JSON encoders with the one defined in the parent ones, thus the result would be:
```py
print(B().json())
#> {"dt": "encoded_from_A", "timedt": "encoded_from_B"}
```

As an extension, it should still be possible to override the parent ones:
```py
class Bbis(A):
    timedt: timedelta = timedelta(hours=100)

    class Config:
        json_encoders = {
            datetime: lambda _: "encoded_from_Bbis",
            timedelta: lambda _: "encoded_from_Bbis",
        }

print(Bbis().json()) 
#> {"dt": "encoded_from_Bbis", "timedt": "encoded_from_Bbis"}
```

I'm eager to implement this feature if you think it's worth it :smile:

### PR DESCRIPTION
**Pull request by art049**
## Change Summary

While inheriting the configuration , the `json_encoders` dictionaries are merged and the child classes can still override an encoder  defined in the parent class.

## Related issue number

Fixes #2024 

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

File: pydantic/main.py [hunk 1]
```diff
@@ -158,13 +158,20 @@ def prepare_field(cls, field: 'ModelField') -> None:
...
     elif self_config == parent_config:
         base_classes = (self_config,)
     else:
         base_classes = self_config, parent_config  # type: ignore
-    return type('Config', base_classes, {})
+        json_encoders = {
+            **getattr(parent_config, 'json_encoders', {}),
+            **getattr(self_config, 'json_encoders', {}),
+        }
+        namespace['json_encoders'] = json_encoders
```
Comment (samuelcolvin) [inline pydantic/main.py:172]:
```suggestion
        namespace['json_encoders'] = {
            **getattr(parent_config, 'json_encoders', {}),
            **getattr(self_config, 'json_encoders', {}),
        }
```

**Review by samuelcolvin — COMMENTED** [review]
you need to add a change description and update the docs to describe the new behaviour preferably with an example

**Comment by art049** [comment]
> you need to add a change description and update the docs to describe the new behaviour preferably with an example
Thanks for the feedback @samuelcolvin  !
I wasn't very sure this feature was ok as I didn't have any feedback neither on the issue nor the PR so I didn't complete it but as it seems ok to you I will be happy complete the changes 😄

**Comment by samuelcolvin** [comment]
@art049, I'm happy to accept this change.

We still need a change description and docs.

**Comment by samuelcolvin** [comment]
@art049 are you okay to fix this?

**Comment by art049** [comment]
> @art049 are you okay to fix this?
@samuelcolvin yes, totally.
I fixed the code smell and added the required documentation :smile: .

WDYT about the docs? I don't know if it explains the feature properly enough

**Comment by samuelcolvin** [comment]
great, thank you so much.

**Pull request merged**